kong/runloop/balancer/init.lua

```diff
@@ -346,6 +346,15 @@ local function set_host_header(balancer_data)
     upstream_host = upstream_host .. ":" .. balancer_data.port
   end
 
+  -- the nginx grpc module does not offer a way to override the
+  -- :authority pseudo-header; use our internal API to do so
+  if upstream_scheme == "grpc" or upstream_scheme == "grpcs" then
+    local ok, err = kong.service.request.set_header(":authority", upstream_host)
```

Comment:
When I read the PDK `set_header` function, it seems only thing this does is:
```
require("resty.kong.grpc").set_authority(upstream_host)
```

Should we skip PDK in this hot path?

Reply:
Good point. I think we had a similar discussion originally but ended up opting to move this into the PDK. @dndx do you have any thoughts here? I am fine either way.

Reply:
I think we said in the past that it is generally advisable to use PDK functions in core code if possible as it makes a good example for other developers to rely more on PDK.

In practice, this codepath should be optimized by the JIT compiler anyway so the actual runtime penalty should be minimum. It is probably not worth making into the `set_authority` function in this case.

If we _really_ want, we can cache the `kong.service.request.set_header` function. But again, probably not worth it in this cas.